azure-iot-device/azure/iot/device/provisioning/sk_registration_client.py

```diff
@@ -0,0 +1,80 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""
+This module contains one of the implementations of the Registration Client which uses Symmetric Key authentication.
+"""
+import logging
+from threading import Event
+from .registration_client import RegistrationClient
+from .models.registration_result import RegistrationResult
+from .internal.polling_machine import PollingMachine
+
+logger = logging.getLogger(__name__)
+
+
+class SymmetricKeyRegistrationClient(RegistrationClient):
+    """
+    Client which can be used to run the registration of a device using Symmetric Key authentication.
+    """
+
+    def __init__(self, mqtt_state_based_provider):
+        """
+        Initializer for the Symmetric Key Registration Client
+        """
+        super(SymmetricKeyRegistrationClient, self).__init__(mqtt_state_based_provider)
+        self._polling_machine = PollingMachine(mqtt_state_based_provider)
+        # To be defined by sample
+        self.on_registration_update = None
+        self._polling_machine.on_connected = self._on_state_change
+        self._polling_machine.on_disconnected = self._on_state_change
+        self._polling_machine.on_registration_status_update = self._on_device_registration_update
+
+    def register(self):
+        """
+        Register the device with the provisioning service.
+        """
+        logger.info("Registering with Hub...")
+        register_complete = Event()
+
+        def callback_register(result):
+            register_complete.set()
+            # TODO This could be a failed result or a successful one.
+            # TODO Response should be given appropriately
+            if isinstance(result, RegistrationResult):
```

Comment:
>if isinstance(result, RegistrationResult): [](start = 12, length = 42)

I get very nervous around isinstance calls. They're usually avoidable, and their presence generally indicates something a little odd.

In this case, I think we're doing the check because this result field could be, in addition to a RegistrationResult, also an error? And we're just passing either in on the "result" parameter? That seems not a usage of the duck typing concept, but an abuse of it. These two possible types of values share nothing in common as actual objects that would allow for them to both "quack like a duck", they're totally separate objects which happen to both come from the same place.

Reply:
taking 2 parameters instead of 1 

---
In reply to: [277779840](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277779840) [](ancestors = 277779840)